frame/staking-rewards/src/lib.rs

```diff
@@ -599,8 +620,119 @@ pub mod pallet {
+				// erroring, short-circuiting on any error found and returning the last known
+				// "good" value.
+				let new_total_rewards = iter::repeat(reward.reward_rate.amount)
+					// REVIEW(benluelo): this cast is *probably* safe, but
+					// should be verified somehow
+					.take(periods_surpassed as usize)
+					.try_fold(reward.total_rewards, |current_total_amount, reward_amount| {
+						match current_total_amount.safe_add(&reward_amount) {
+							Ok(new_total_amount) if new_total_amount <= reward.max_rewards =>
+								ControlFlow::Continue(new_total_amount),
+							// max rewards hit or overflow
+							// REVIEW(benluelo): Partial rewards accumulation?
+							// i.e. max rewards is 100, amount is 30, period is 1 second; after 4
+							// seconds the total accumulated will be 120, but that's more than
+							// the max amount so it gets reverted to 90.
+							Ok(_) | Err(_) => {
+								Self::deposit_event(Event::<T>::RewardAccumulationError {
+									pool_id,
+									asset_id,
+								});
+								ControlFlow::Break(current_total_amount)
+							},
+						}
+					});
+
+				(
+					asset_id,
+					Reward {
+						total_rewards: match new_total_rewards {
+							ControlFlow::Continue(new_total_rewards) => new_total_rewards,
+							ControlFlow::Break(new_total_rewards) => new_total_rewards,
+						},
+						last_updated_timestamp: now,
+						..reward
+					},
+				)
+			}
+		}
+
+		// TODO: tests
+		pub(crate) fn acumulate_rewards_hook() -> Weight {
+			let now = T::UnixTime::now().as_secs();
+
+			RewardPools::<T>::iter()
+				.into_iter()
+				.map(|(pool_id, reward_pool)| {
+					let updated_rewards = reward_pool
+						.rewards
+						.into_iter()
+						.map(|(asset_id, reward)| {
+							Self::reward_acumulation_hook_reward_update_calculation(
+								pool_id, asset_id, reward, now,
+							)
+						})
+						.try_collect()
+						// SAFETY(benluelo): No elements were added to the BTreeMap; the only
+						// operation was `.map()`. This expect call will be unnecessary once this is
+						// merged and we update to whatever version it's included in:
+						// https://github.com/paritytech/substrate/pull/11869
+						.expect("no elements were added; qed");
+
+					(pool_id, RewardPool { rewards: updated_rewards, ..reward_pool })
+				})
+				// NOTE(benluelo): As per these docs on `StorageMap::iter`:
+				// https://github.com/paritytech/substrate/blob/cac91f59b9e3fb8fd59842c023f87b4206931993/frame/support/src/storage/types/map.rs,
+				// "If you alter the map while doing this, you'll get undefined results."
+				// hence the double iteration.
+				.collect::<Vec<_>>()
+				.into_iter()
+				.fold(T::WeightInfo::unix_time_now(), |total_weights, (pool_id, reward_pool)| {
+					// 128 bit platforms don't exist as of writing this so this cast should be ok
+					let amount_of_rewards_in_pool = reward_pool.rewards.len() as u64;
+
+					RewardPools::<T>::insert(pool_id, reward_pool);
+
+					total_weights +
```

Comment:
IMO this code would be much more readable if we did the weight updates in a for or while loop with the actual logic.

Reply:
This got buried, I didn't see it. See https://github.com/ComposableFi/composable/pull/1359/commits/e9cc52cfc36892769e1a8a697bd77557559d7938